### ISSUE TITLE
Run OpenDevin in Docker container

### ISSUE THREAD

**Issue by grewhit25** [OpenDevin/OpenDevin#1104]
<!-- You MUST fill out this template. We will close issues that don't include enough information to reproduce -->
#### Describe the bug
<!-- a short description of the problem -->

Docker build process: "pull-docker-image" and "check-docker" fails within container.

description:
I'm attempting to run OpenDevin inside a Docker container, built using the official build process. While everything seems to be working well - the Sandbox container pulls and starts automatically with the latest updates - I've encountered an issue with the make build process within the container.

Current Behavior:
The pull-docker-image and check-docker dependencies in the Makefile don't work correctly when running the build process inside the Docker container.

Workaround:
As a temporary solution, I've added the following lines to the Dockerfile to skip these checks:
Code snippet:

```bash

sed -i '/@$(MAKE) -s pull-docker-image/d' Makefile
sed -i '/@$(MAKE) -s check-docker/d' Makefile
```

Proposed Improvement:
To address this issue more elegantly, we could potentially introduce a flag in the Makefile:
Code snippet:

```bash

RUN_IN_DOCKER = true/false

if [[ "$RUN_IN_DOCKER" == "true" ]]; then
  echo "Running in Docker"
  # skip docker checks
else
  echo "Not running in Docker"
  # run docker checks
fi
```

This allows the build process to determine if it's running inside a container and skip the problematic steps when necessary.

Additional Information:
I've included a sample Dockerfile, run script, and startup logs for reference.

Request:
I'd appreciate some guidance on how to best address this issue. Is the proposed Makefile logic a viable approach, or are there alternative solutions available?

#### Setup and configuration
**Current version**:
<!-- run `git log -n 1` to see this -->

```bash

commit 2326312d3a297a5027b83018c0a5e0421ff74285 (HEAD -> main, origin/main, origin/HEAD)
Author: Leo <ifuryst@gmail.com>
Date:   Sat Apr 13 10:18:34 2024 +0800

    fix: print the wrong ssh port number (#1054)

```

<!-- tell us everything about your environment -->
**My config.toml and environment vars** (be sure to redact API keys):

```toml

# This is config.toml`.
LLM_MODEL="gpt-4-0125-preview"
LLM_API_KEY=""

WORKSPACE_DIR="./workspace"
SANDBOX_TYPE='exec'

```

**My model and agent** (you can see these settings in the UI):
* Model:
* Agent:

**Commands I ran to install and run OpenDevin**:

```bash

# Install OpenDevin
cd myworkingdirectory
git clone https://github.com/OpenDevin/OpenDevin.git

# With the sample Dockerfile
docker build -t localhost/opendevin-poetry .

# Run OpenDevin
# with the sample run script
./run_opendevin.sh

```

**Steps to Reproduce**:
1.
2.
3.

**Logs, error messages, and screenshots**:

#### Additional Context

<details>
<summary>Sample Dockerfile</summary>
<br>

```dockerfile

FROM ubuntu:22.04 as base

ARG DEBIAN_FRONTEND=noninteractive
ARG S6_OVERLAY_VERSION=3.1.6.2

# Install base packages
RUN apt update && apt install -y software-properties-common \
    && apt update &&  add-apt-repository ppa:deadsnakes/ppa -y \
    && apt update && apt install -y \    
    python3.11 python3-pip python3-venv python3-dev \
    curl wget git \
    && curl -fsSL https://deb.nodesource.com/setup_20.x | bash - \
    && apt install -y nodejs \
    && update-alternatives --install /usr/bin/python python /usr/bin/python3.11 1 \
    # https://python-poetry.org/docs#installing-with-the-official-installer then add poetry to the PATH
    && curl -sSL https://install.python-poetry.org | python - \
    && apt clean \
    && rm -rf /var/lib/apt/lists/*


# Install s6-overlay
ADD https://github.com/just-containers/s6-overlay/releases/download/v${S6_OVERLAY_VERSION}/s6-overlay-noarch.tar.xz /tmp
RUN tar -C / -Jxpf /tmp/s6-overlay-noarch.tar.xz
ADD https://github.com/just-containers/s6-overlay/releases/download/v${S6_OVERLAY_VERSION}/s6-overlay-x86_64.tar.xz /tmp
RUN tar -C / -Jxpf /tmp/s6-overlay-x86_64.tar.xz

# Install OpenDevin
FROM base as app
# Configur default environment variables
ENV DOCKER_IMAGE ghcr.io/opendevin/sandbox
ENV BACKEND_PORT 3000
ENV BACKEND_HOST "127.0.0.1:$BACKEND_PORT"
ENV FRONTEND_PORT 3001
ENV BDEFAULT_WORKSPACE_DIR "./workspace"
ENV DEFAULT_MODEL "gpt-3.5-turbo-1106"
ENV CONFIG_FILE = config.toml
ENV PATH="/root/.local/bin:$PATH"

# Clone the OpenDevin repository and build OpenDevin, note this is a moving target
# RUN git clone https://github.com/OpenDevin/OpenDevin.git /app/opendevin

# Or use a local copy of the repository
COPY OpenDevin /app/opendevin
WORKDIR /app/opendevin

# Use the official process to build OpenDevin, however you have to
# Workaround: delete these lines from the Makefile so that make build doesn't fail on those checks.
RUN sed -i '/@$(MAKE) -s pull-docker-image/d' Makefile \
    && sed -i '/@$(MAKE) -s check-docker/d' Makefile
    
RUN make build

# COPY s6 /etc/s6-overlay/s6-rc.d/
COPY entrypoint.sh /app
RUN chmod +x /app/entrypoint.sh

# Temporary work around for UID 0 issue https://github.com/OpenDevin/OpenDevin/issues/936
ENV SANDBOX_USER_ID=1000

# ENTRYPOINT ["/init"]
ENTRYPOINT ["/app/entrypoint.sh"]

```
</details>

<details>
<summary>Sample entrypoint.sh</summary>
<br>

```bash

#!/bin/bash
set -euo pipefail

# run uvicorn opendevin.server.listen:app --port 3000 and npm start from here
# find / -name 'libcudart.so'

cd /app/opendevin
poetry run uvicorn opendevin.server.listen:app --reload --port $BACKEND_PORT &
cd frontend
BACKEND_PORT=$BACKEND_PORT
FRONTEND_PORT=$FRONTEND_PORT
BACKEND_HOST=$BACKEND_HOST
npm run start -- --host $BACKEND_HOST

exec "$@"

```

</details>

<details>
<summary>Sample run script</summary>
<br>

```bash

# !/bin/bash
# run the opendevin server
set -euo pipefail
IFS=$'\n\t'
args=(
    # -p 3000:3000
    -p 3001:3001
    --name opendevin
#    --env-file ~/apps/opendevin/.env
    -v /var/run/docker.sock:/var/run/docker.sock:ro
    -v ./config.toml:/app/opendevin/config.toml:ro
    -v /app/opendevin/workspace:/app/opendevin/workspace
    -e BACKEND_HOST=0.0.0.0
   --restart=unless-stopped
)
if [ "$(docker ps -aq -f name=opendevin)" ]; then
    docker stop opendevin
    docker rm opendevin
fi
docker run  -d "${args[@]}" localhost/opendevin-poetry

```

Note: the image is build locally in this example as opendevin-poetry
</details>

<details>
<summary>Sample log output</summary>
<br>

```log

docker logs opendevin                                                                                                            
s6-rc: info: service backend: starting                                                                                             s6-rc: info: service s6rc-oneshot-runner: starting                                                                                 
s6-rc: info: service backend successfully started                                                                                  
s6-rc: info: service s6rc-oneshot-runner successfully started                                                                      
s6-rc: info: service fix-attrs: starting                                                                                           
Started backend server... on port 3000                                                                                             s6-rc: info: service fix-attrs successfully started                                                                                
s6-rc: info: service legacy-cont-init: starting                                                                                    
s6-rc: info: service legacy-cont-init successfully started                                                                         
s6-rc: info: service frontend: starting                                                                                            
s6-rc: info: service frontend successfully started                                                                                 s6-rc: info: service legacy-services: starting                                                                                     
s6-rc: info: service legacy-services successfully started                                                                          
                                                                                                                                   
> opendevin-frontend@0.1.0 start                                                                                                   
> vite --host 0.0.0.0                                                                                                              
                                                                                                                                   
INFO:     Will watch for changes in these directories: ['/app/opendevin']                                                          
INFO:     Uvicorn running on http://0.0.0.0:3000 (Press CTRL+C to quit)                                                            
INFO:     Started reloader process [43] using WatchFiles                                                                           
                                                                                                                                   
  VITE v5.2.8  ready in 331 ms                                                                                                     
                                                                                                                                   
  ➜  Local:   http://localhost:3001/                                                                                               
  ➜  Network: http://172.17.0.3:3001/                                                                                              
/root/.cache/pypoetry/virtualenvs/opendevin-Bs3vTTT_-py3.11/lib/python3.11/site-packages/pydantic/_internal/_fields.py:151: UserWar
ning: Field "model_list" has conflict with protected namespace "model_".  

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.                            
  warnings.warn(
/root/.cache/pypoetry/virtualenvs/opendevin-Bs3vTTT_-py3.11/lib/python3.11/site-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_group_alias" has conflict with protected namespace "model_".                                            

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.                            
  warnings.warn(
/root/.cache/pypoetry/virtualenvs/opendevin-Bs3vTTT_-py3.11/lib/python3.11/site-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_info" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.                            
  warnings.warn(
/root/.cache/pypoetry/virtualenvs/opendevin-Bs3vTTT_-py3.11/lib/python3.11/site-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_id" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.                            
  warnings.warn(
INFO:     Started server process [105]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     ('127.0.0.1', 53222) - "WebSocket /ws?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzaWQiOiJlNmY1MmJhOS0zMjZjLTQ0Y2YtYTRjYS0yODhkZjc3MWM4ODgifQ.2GpcPc5TpOH_FPT5lAEV7cOJg9xbct76thuhFyuC8kg" [accepted]
INFO:     connection open
Starting loop_recv for sid: e6f52ba9-326c-44cf-a4ca-288df771c888
INFO:     127.0.0.1:53236 - "GET /refresh-files HTTP/1.1" 200 OK
INFO:     127.0.0.1:53238 - "GET /configurations HTTP/1.1" 200 OK
INFO:     127.0.0.1:53250 - "GET /messages/total HTTP/1.1" 200 OK
INFO:     127.0.0.1:53266 - "GET /refresh-files HTTP/1.1" 200 OK
08:13:24 - opendevin:INFO: exec_box.py:75 - Using workspace directory: /app/opendevin/workspace
08:13:24 - opendevin:INFO: exec_box.py:198 - Container stopped
08:13:24 - opendevin:INFO: exec_box.py:215 - Container started

```

<!-- You MUST fill out this template. We will close issues that don't include enough information to reproduce -->
#### Describe the bug
<!-- a short description of the problem -->
description:
I'm attempting to run OpenDevin inside a Docker container, built using the official build process. While everything seems to be working well - the Sandbox container pulls and starts automatically with the latest updates - I've encountered an issue with the make build process within the container.

Docker build process: "pull-docker-image" and "check-docker" fails within container.

Current Behavior:
The pull-docker-image and check-docker dependencies in the Makefile don't work correctly when running the build process inside the Docker container.

Workaround:
As a temporary solution, I've added the following lines to the Dockerfile to skip these checks:
Code snippet:

```bash

sed -i '/@$(MAKE) -s pull-docker-image/d' Makefile
sed -i '/@$(MAKE) -s check-docker/d' Makefile
```

Proposed Improvement:
To address this issue more elegantly, we could potentially introduce a flag in the Makefile:
Code snippet:

```bash

RUN_IN_DOCKER = true/false

if [[ "$RUN_IN_DOCKER" == "true" ]]; then
  echo "Running in Docker"
  # skip docker checks
else
  echo "Not running in Docker"
  # run docker checks
fi
```

This allows the build process to determine if it's running inside a container and skip the problematic steps when necessary.

Additional Information:
I've included a sample Dockerfile, run script, and startup logs for reference.

Request:
I'd appreciate some guidance on how to best address this issue. Is the proposed Makefile logic a viable approach, or are there alternative solutions available?

#### Setup and configuration
**Current version**:
<!-- run `git log -n 1` to see this -->

```bash

commit 2326312d3a297a5027b83018c0a5e0421ff74285 (HEAD -> main, origin/main, origin/HEAD)
Author: Leo <ifuryst@gmail.com>
Date:   Sat Apr 13 10:18:34 2024 +0800

    fix: print the wrong ssh port number (#1054)

```

<!-- tell us everything about your environment -->
**My config.toml and environment vars** (be sure to redact API keys):

```toml

# This is config.toml`.
LLM_MODEL="gpt-4-0125-preview"
LLM_API_KEY=""

WORKSPACE_DIR="./workspace"
SANDBOX_TYPE='exec'

```

**My model and agent** (you can see these settings in the UI):
* Model:
* Agent:

**Commands I ran to install and run OpenDevin**:

```bash

# Install OpenDevin
cd myworkingdirectory
git clone https://github.com/OpenDevin/OpenDevin.git

# With the sample Dockerfile
docker build -t localhost/opendevin-poetry .

# Run OpenDevin
# with the sample run script
./run_opendevin.sh

```

**Steps to Reproduce**:
1.
2.
3.

**Logs, error messages, and screenshots**:

#### Additional Context

<details>
<summary>Sample Dockerfile</summary>
<br>

```dockerfile

FROM ubuntu:22.04 as base

ARG DEBIAN_FRONTEND=noninteractive
ARG S6_OVERLAY_VERSION=3.1.6.2

# Install base packages
RUN apt update && apt install -y software-properties-common \
    && apt update &&  add-apt-repository ppa:deadsnakes/ppa -y \
    && apt update && apt install -y \    
    python3.11 python3-pip python3-venv python3-dev \
    curl wget git \
    && curl -fsSL https://deb.nodesource.com/setup_20.x | bash - \
    && apt install -y nodejs \
    && update-alternatives --install /usr/bin/python python /usr/bin/python3.11 1 \
    # https://python-poetry.org/docs#installing-with-the-official-installer then add poetry to the PATH
    && curl -sSL https://install.python-poetry.org | python - \
    && apt clean \
    && rm -rf /var/lib/apt/lists/*


# Install s6-overlay
ADD https://github.com/just-containers/s6-overlay/releases/download/v${S6_OVERLAY_VERSION}/s6-overlay-noarch.tar.xz /tmp
RUN tar -C / -Jxpf /tmp/s6-overlay-noarch.tar.xz
ADD https://github.com/just-containers/s6-overlay/releases/download/v${S6_OVERLAY_VERSION}/s6-overlay-x86_64.tar.xz /tmp
RUN tar -C / -Jxpf /tmp/s6-overlay-x86_64.tar.xz

# Install OpenDevin
FROM base as app
# Configur default environment variables
ENV DOCKER_IMAGE ghcr.io/opendevin/sandbox
ENV BACKEND_PORT 3000
ENV BACKEND_HOST "127.0.0.1:$BACKEND_PORT"
ENV FRONTEND_PORT 3001
ENV BDEFAULT_WORKSPACE_DIR "./workspace"
ENV DEFAULT_MODEL "gpt-3.5-turbo-1106"
ENV CONFIG_FILE = config.toml
ENV PATH="/root/.local/bin:$PATH"

# Clone the OpenDevin repository and build OpenDevin, note this is a moving target
# RUN git clone https://github.com/OpenDevin/OpenDevin.git /app/opendevin

# Or use a local copy of the repository
COPY OpenDevin /app/opendevin
WORKDIR /app/opendevin

# Use the official process to build OpenDevin, however you have to
# Workaround: delete these lines from the Makefile so that make build doesn't fail on those checks.
RUN sed -i '/@$(MAKE) -s pull-docker-image/d' Makefile \
    && sed -i '/@$(MAKE) -s check-docker/d' Makefile
    
RUN make build

# COPY s6 /etc/s6-overlay/s6-rc.d/
COPY entrypoint.sh /app
RUN chmod +x /app/entrypoint.sh

# Temporary work around for UID 0 issue https://github.com/OpenDevin/OpenDevin/issues/936
ENV SANDBOX_USER_ID=1000

# ENTRYPOINT ["/init"]
ENTRYPOINT ["/app/entrypoint.sh"]

```
</details>

<details>
<summary>Sample entrypoint.sh</summary>
<br>

```bash

#!/bin/bash
set -euo pipefail

# run uvicorn opendevin.server.listen:app --port 3000 and npm start from here
# find / -name 'libcudart.so'

cd /app/opendevin
poetry run uvicorn opendevin.server.listen:app --reload --port $BACKEND_PORT &
cd frontend
BACKEND_PORT=$BACKEND_PORT
FRONTEND_PORT=$FRONTEND_PORT
BACKEND_HOST=$BACKEND_HOST
npm run start -- --host $BACKEND_HOST

exec "$@"

```

</details>

<details>
<summary>Sample run script</summary>
<br>

```bash

# !/bin/bash
# run the opendevin server
set -euo pipefail
IFS=$'\n\t'
args=(
    # -p 3000:3000
    -p 3001:3001
    --name opendevin
#    --env-file ~/apps/opendevin/.env
    -v /var/run/docker.sock:/var/run/docker.sock:ro
    -v ./config.toml:/app/opendevin/config.toml:ro
    -v /app/opendevin/workspace:/app/opendevin/workspace
    -e BACKEND_HOST=0.0.0.0
   --restart=unless-stopped
)
if [ "$(docker ps -aq -f name=opendevin)" ]; then
    docker stop opendevin
    docker rm opendevin
fi
docker run  -d "${args[@]}" localhost/opendevin-poetry

```

Note: the image is build locally in this example as opendevin-poetry
</details>

<details>
<summary>Sample log output</summary>
<br>

```log

docker logs opendevin                                                                                                            
s6-rc: info: service backend: starting                                                                                             s6-rc: info: service s6rc-oneshot-runner: starting                                                                                 
s6-rc: info: service backend successfully started                                                                                  
s6-rc: info: service s6rc-oneshot-runner successfully started                                                                      
s6-rc: info: service fix-attrs: starting                                                                                           
Started backend server... on port 3000                                                                                             s6-rc: info: service fix-attrs successfully started                                                                                
s6-rc: info: service legacy-cont-init: starting                                                                                    
s6-rc: info: service legacy-cont-init successfully started                                                                         
s6-rc: info: service frontend: starting                                                                                            
s6-rc: info: service frontend successfully started                                                                                 s6-rc: info: service legacy-services: starting                                                                                     
s6-rc: info: service legacy-services successfully started                                                                          
                                                                                                                                   
> opendevin-frontend@0.1.0 start                                                                                                   
> vite --host 0.0.0.0                                                                                                              
                                                                                                                                   
INFO:     Will watch for changes in these directories: ['/app/opendevin']                                                          
INFO:     Uvicorn running on http://0.0.0.0:3000 (Press CTRL+C to quit)                                                            
INFO:     Started reloader process [43] using WatchFiles                                                                           
                                                                                                                                   
  VITE v5.2.8  ready in 331 ms                                                                                                     
                                                                                                                                   
  ➜  Local:   http://localhost:3001/                                                                                               
  ➜  Network: http://172.17.0.3:3001/                                                                                              
/root/.cache/pypoetry/virtualenvs/opendevin-Bs3vTTT_-py3.11/lib/python3.11/site-packages/pydantic/_internal/_fields.py:151: UserWar
ning: Field "model_list" has conflict with protected namespace "model_".  

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.                            
  warnings.warn(
/root/.cache/pypoetry/virtualenvs/opendevin-Bs3vTTT_-py3.11/lib/python3.11/site-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_group_alias" has conflict with protected namespace "model_".                                            

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.                            
  warnings.warn(
/root/.cache/pypoetry/virtualenvs/opendevin-Bs3vTTT_-py3.11/lib/python3.11/site-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_info" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.                            
  warnings.warn(
/root/.cache/pypoetry/virtualenvs/opendevin-Bs3vTTT_-py3.11/lib/python3.11/site-packages/pydantic/_internal/_fields.py:151: UserWarning: Field "model_id" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.                            
  warnings.warn(
INFO:     Started server process [105]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     ('127.0.0.1', 53222) - "WebSocket /ws?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzaWQiOiJlNmY1MmJhOS0zMjZjLTQ0Y2YtYTRjYS0yODhkZjc3MWM4ODgifQ.2GpcPc5TpOH_FPT5lAEV7cOJg9xbct76thuhFyuC8kg" [accepted]
INFO:     connection open
Starting loop_recv for sid: e6f52ba9-326c-44cf-a4ca-288df771c888
INFO:     127.0.0.1:53236 - "GET /refresh-files HTTP/1.1" 200 OK
INFO:     127.0.0.1:53238 - "GET /configurations HTTP/1.1" 200 OK
INFO:     127.0.0.1:53250 - "GET /messages/total HTTP/1.1" 200 OK
INFO:     127.0.0.1:53266 - "GET /refresh-files HTTP/1.1" 200 OK
08:13:24 - opendevin:INFO: exec_box.py:75 - Using workspace directory: /app/opendevin/workspace
08:13:24 - opendevin:INFO: exec_box.py:198 - Container stopped
08:13:24 - opendevin:INFO: exec_box.py:215 - Container started

```

**Comment by loopyd** [comment]
This is due to a docker buildkit restriction that prevents docker from running within the container without acceptable priviledge escalation at build time.  

Feature from PR #1065 proposes a solution similar to the build flag you mentioned.  

Docker-in-docker to have this feature completely working is on the road map/in progress with multistage / live container / commit to image building proposed by some contributors. Getting it fully working needs work still.

**Comment by grewhit25** [comment]
That’s great thanks for the update.

**Comment by grewhit25** [comment]
Since there is now an official Docker image available, I don’t see the point in this issue remaining open.

I have ran the official image but received connection issue at the time of writing.

It seems the problem is with ‘host.docker.internal’ and SSH_HOSTNAME not being set.
which I have temporary worked around by setting it to my Docker host like in this example:

```bash
-e SSH_HOSTNAME=$(ip route get 8.8.8.8 | head -1 | awk '{print $7}') -e SSH_HOSTNAME=$(ip route get 8.8.8.8 | head -1 | awk '{print $7}')
```

Closing.
